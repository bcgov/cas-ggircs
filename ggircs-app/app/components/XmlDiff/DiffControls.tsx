import React, { useState } from "react";
import { Row, Col } from "react-bootstrap";
import { FontAwesomeIcon } from "@fortawesome/react-fontawesome";
import { faExchangeAlt } from "@fortawesome/free-solid-svg-icons";
import Button from "@button-inc/bcgov-theme/Button";
import { NextRouter } from "next/router";
// Use dynamic import for BootstrapSwitchButton (workaround for SSR bug where window is not defined)
import dynamic from "next/dynamic";
const BootstrapSwitchButton = dynamic(
  () => import("bootstrap-switch-button-react"),
  {
    loading: () => <p>loading</p>,
    ssr: false, // This line is important. It's what prevents server-side render
  }
);

interface Props {
  router: NextRouter;
}

export const DiffDetails: React.FunctionComponent<Props> = ({ router }) => {
  const [collapsed, setCollapsed] = useState(false);
  const [reversed, setReversed] = useState(false);

  const handleRouter = (control: string, value: boolean) => {
    const query = {
      ...router.query,
      [control]: value,
    };
    if (!value) delete query[control];
    const url = {
      pathname: router.pathname,
      query,
    };
    router.push(url, url, { shallow: true });
  };

  const handleCollapse = () => {
    setCollapsed(!collapsed);
    handleRouter("collapsed", !collapsed);
  };

  const handleReverse = () => {
    setReversed(!reversed);
    handleRouter("reversed", !reversed);
  };

  return (
    <Row style={{ marginTop: "2em", marginBottom: "2em" }}>
      <Col md={{ span: 3, offset: 6 }}>
        <Row style={{ float: "right" }}>
          <p style={{ marginTop: "auto", marginBottom: "auto" }}>
            <strong>Summarize Changes:&nbsp;</strong>
          </p>
          <BootstrapSwitchButton
            checked={collapsed}
            onlabel="On"
            offlabel="Off"
            onChange={handleCollapse}
          />
        </Row>
      </Col>
      <Col md={{ span: 3 }}>
        <Button size="small" onClick={handleReverse}>
          <FontAwesomeIcon icon={faExchangeAlt} />
          &nbsp;Swap left/right
        </Button>
      </Col>
    </Row>
  );
};
export default DiffDetails;
