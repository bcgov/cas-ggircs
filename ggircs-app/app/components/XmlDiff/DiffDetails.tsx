import React from "react";
import { Row, Col } from "react-bootstrap";

interface Props {
  zipFileName: string;
  xmlFileName: string;
  submissionDate: string;
}

export const DiffDetails: React.FunctionComponent<Props> = ({
  zipFileName,
  xmlFileName,
  submissionDate,
}) => (
  <Row>
    <Col md={5}>
      <strong>From Archive: </strong>
    </Col>
    <Col md={7}>{zipFileName}</Col>
    <Col md={5}>
      <strong>Name: </strong>
    </Col>
    <Col md={7}>{xmlFileName}</Col>
    <Col md={5}>
      <strong>Date Submitted: </strong>
    </Col>
    <Col md={7}>{submissionDate.split("T")[0]}</Col>
  </Row>
);
export default DiffDetails;
