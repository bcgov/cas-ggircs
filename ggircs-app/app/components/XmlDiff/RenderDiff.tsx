import { useState, useEffect, useMemo } from "react";
import { useFragment, graphql } from "react-relay";
import { diffLines, formatLines } from "unidiff";
import { parseDiff, Diff, Hunk, tokenize } from "react-diff-view";
import "react-diff-view/style/index.css";
import format from "xml-formatter";
import { useRouter } from "next/router";
import LoadingSpinner from "components/LoadingSpinner";
import { RenderDiff_query$key } from "__generated__/RenderDiff_query.graphql";

const EMPTY_HUNKS = [];

interface Props {
  query: RenderDiff_query$key;
}

const RenderDiff: React.FC<Props> = ({
  query
}) => {

  const router = useRouter();
  const {
    firstSideReport,
    secondSideReport
  } = useFragment(
    graphql`
      fragment RenderDiff_query on Query
        @argumentDefinitions(FirstSideRelayId: {type: "ID!"}, SecondSideRelayId: {type: "ID!"}) {
        firstSideReport: report(id: $FirstSideRelayId) {
          swrsReportId
          latestSwrsReport {
            submissionDate
            ecccXmlFileByEcccXmlFileId {
              xmlFileName
              xmlFile
              ecccZipFileByZipFileId {
                zipFileName
              }
            }
          }
        }
        secondSideReport: report(id: $SecondSideRelayId) {
          swrsReportId
          latestSwrsReport {
            submissionDate
            ecccXmlFileByEcccXmlFileId {
              xmlFileName
              xmlFile
              ecccZipFileByZipFileId {
                zipFileName
              }
            }
          }
        }
      }
    `,
    query
  );

  const shouldRenderDiff =
    firstSideReport &&
    secondSideReport &&
    router.query.FirstSideRelayId &&
    router.query.SecondSideRelayId;
  if (!shouldRenderDiff) return null;

  const [isLoading, setIsLoading] = useState(true);
  const [diffText, setDiffText] = useState(null);

  const oldText = format(
    firstSideReport.latestSwrsReport.ecccXmlFileByEcccXmlFileId.xmlFile
  );
  const newText = format(
    secondSideReport.latestSwrsReport.ecccXmlFileByEcccXmlFileId.xmlFile
  );

  useEffect(() => {
    setIsLoading(false);
  });

  useMemo(() => {
    setIsLoading(true);
    if (router.query.reversed)
      setDiffText(
        formatLines(diffLines(newText, oldText), {
          // context is the max number of lines to show around a change. When collapsed, we set it to 1. Otherwise 10000 to ensure we get the whole document.
          context: router.query.collapsed ? 1 : 10000,
        })
      );
    else
      setDiffText(
        formatLines(diffLines(oldText, newText), {
          context: router.query.collapsed ? 1 : 10000,
        })
      );
  }, [router.query]);

  if (isLoading) return <LoadingSpinner />;

  const [{ type, hunks }] = parseDiff(diffText, { nearbySequences: "zip" });

  const getTokens = (hunks) => {
    if (!hunks) {
      return undefined;
    }

    const options = {
      highlight: false,
      language: "markup",
    };

    try {
      return tokenize(hunks, options);
    } catch (ex) {
      return undefined;
    }
  };

  const tokens = getTokens(hunks);

  const renderToken = (token, defaultRender, i) => {
    const pttrn = /^\s*</;
    const str = token.value;
    const leadingWhiteSpace = str.match(pttrn)?.[0].length - 1;
    let tagColor = "black";

    const tagHighlighterArray = [
      "#0000e0",
      "#aa0000",
      "#ff4500",
      "#360036",
      "#ff00ff",
      "#8d6708",
      "#1978d4",
    ];
    if (Number.isInteger(leadingWhiteSpace))
      leadingWhiteSpace < 25
        ? (tagColor = tagHighlighterArray[leadingWhiteSpace / 4])
        : "#ff0000";


    if (token.type === "text") {
      return (
        <span key={i} style={{ color: tagColor }}>
          {token.value}
        </span>
      );
    }

    // For other types, use the default render function
    return defaultRender(token, i);
  };

  return (
    <div style={{ height: "40em", overflow: "scroll" }}>
      <Diff
        viewType="split"
        diffType={type}
        hunks={hunks || EMPTY_HUNKS}
        tokens={tokens}
        renderToken={renderToken}
      >
        {(hunks) =>
          hunks.map((hunk) => <Hunk key={hunk.content} hunk={hunk} />)
        }
      </Diff>
    </div>
  );
};

export default RenderDiff;
