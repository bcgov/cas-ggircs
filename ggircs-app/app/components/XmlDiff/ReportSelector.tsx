import { useCallback, useState } from "react";
import { RelayReportObject } from "types";
import Input from "@button-inc/bcgov-theme/Input";
import { FontAwesomeIcon } from "@fortawesome/react-fontawesome";
import { faCheck, faTimes } from "@fortawesome/free-solid-svg-icons";
import debounce from "lodash.debounce";
import { useRouter } from "next/router";

interface Props {
  diffSide: String;
  allReports: readonly RelayReportObject[];
  onUserTyping: (isUserTyping: boolean) => void;
  disabled: boolean;
}

export const ReportSelector: React.FunctionComponent<Props> = ({
  diffSide,
  allReports,
  onUserTyping,
  disabled,
}) => {
  const router = useRouter();

  const [swrsReportIdIsValid, setSwrsReportIdIsvalid] = useState<boolean>(null);
  const [swrsReportId, setSwrsReportId] = useState<number>(
    Number(router.query[`${diffSide}SideId`])
  );

  const handleRouter = useCallback(
    (id: number, relayId: string, valid: boolean) => {
      let query;
      if (valid) {
        query = {
          ...router.query,
          [`${diffSide}SideId`]: id,
          [`${diffSide}SideRelayId`]: relayId,
        };
      } else {
        query = { ...router.query };
        delete query[`${diffSide}SideId`];
        delete query[`${diffSide}SideRelayId`];
      }
      if (!relayId) delete query[`${diffSide}SideRelayId`];
      const url = {
        pathname: router.pathname,
        query,
      };
      router.push(url, url, { shallow: true });
    },
    [router, diffSide]
  );

  const validateReportId = (id: number) => {
    const edge = allReports.find((edge) => edge?.node.swrsReportId === id);
    if (!id) {
      setSwrsReportIdIsvalid(null);
      handleRouter(null, null, false);
    } else if (edge) {
      setSwrsReportIdIsvalid(true);
      handleRouter(id, edge?.node.id, true);
    } else {
      setSwrsReportIdIsvalid(false);
      handleRouter(id, edge?.node.id, false);
    }
  };

  const debouncedToRouter = useCallback(
    debounce((nextValue) => {
      onUserTyping(false);
      validateReportId(Number(nextValue));
    }, 1000),
    [router]
  );

  const handleChange = (event) => {
    const { value: nextValue } = event.target;
    setSwrsReportId(nextValue);
    onUserTyping(true);
    debouncedToRouter(nextValue);
  };

  const idSelector = (
    <Input
      id="report-id-input"
      type="number"
      placeholder="eg. 17778"
      maxLength={15}
      name="report-id-input"
      rounded
      size="medium"
      onChange={handleChange}
      value={String(swrsReportId)}
      disabled={disabled}
    />
  );

  return (
    <div>
      <div>
        <strong>{diffSide} File for Comparison:</strong>
      </div>
      <p>
        <em>
          Enter the SWRS Report Id. The file contents will show below
          automatically.
        </em>
      </p>
      {idSelector}
      {swrsReportIdIsValid === false && swrsReportId && (
        <small style={{ color: "red" }}>
          <FontAwesomeIcon icon={faTimes} />
          &nbsp;The ID you have entered does not exist
        </small>
      )}
      {swrsReportIdIsValid && swrsReportId && (
        <small style={{ color: "green" }}>
          <FontAwesomeIcon icon={faCheck} />
          &nbsp;ID is valid
        </small>
      )}
    </div>
  );
};

export default ReportSelector;
