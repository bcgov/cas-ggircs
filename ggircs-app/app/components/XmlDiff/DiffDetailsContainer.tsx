import { useFragment, graphql } from "react-relay";
import { Row, Col } from "react-bootstrap";
import { DiffDetailsContainer_query$key } from "DiffDetailsContainer_query.graphql";
import DiffDetails from "components/XmlDiff/DiffDetails";
import { useRouter } from "next/router";

interface Props {
  query: DiffDetailsContainer_query$key;
}

export const DiffDetailsContainer: React.FC<Props> = ({
  query
}) => {

  const router = useRouter();
  const {
    firstSideDetails,
    secondSideDetails
  } = useFragment(
    graphql`
      fragment DiffDetailsContainer_query on Query
        @argumentDefinitions(FirstSideRelayId: {type: "ID!"}, SecondSideRelayId: {type: "ID!"}) {
        firstSideDetails: report(id: $FirstSideRelayId) {
          latestSwrsReport {
            submissionDate
            ecccXmlFileByEcccXmlFileId {
              xmlFileName
              xmlFile
              ecccZipFileByZipFileId {
                zipFileName
              }
            }
          }
        }
        secondSideDetails: report(id: $SecondSideRelayId) {
          latestSwrsReport {
            submissionDate
            ecccXmlFileByEcccXmlFileId {
              xmlFileName
              xmlFile
              ecccZipFileByZipFileId {
                zipFileName
              }
            }
          }
        }
      }
    `,
    query
  );


  return (
    <Row style={{ marginTop: "2em" }}>
      <Col md={{ span: 6, order: router.query.reversed ? 2 : 1 }}>
        {firstSideDetails && router.query.FirstSideRelayId && (
          <DiffDetails
            zipFileName={
              firstSideDetails.latestSwrsReport.ecccXmlFileByEcccXmlFileId
                .ecccZipFileByZipFileId.zipFileName
            }
            xmlFileName={
              firstSideDetails.latestSwrsReport.ecccXmlFileByEcccXmlFileId
                .xmlFileName
            }
            submissionDate={
              firstSideDetails.latestSwrsReport.submissionDate
            }
          />
        )}
      </Col>
      <Col md={{ span: 6, order: router.query.reversed ? 1 : 2 }}>
        {secondSideDetails && router.query.SecondSideRelayId && (
          <DiffDetails
            zipFileName={
              secondSideDetails.latestSwrsReport.ecccXmlFileByEcccXmlFileId
                .ecccZipFileByZipFileId.zipFileName
            }
            xmlFileName={
              secondSideDetails.latestSwrsReport.ecccXmlFileByEcccXmlFileId
                .xmlFileName
            }
            submissionDate={
              secondSideDetails.latestSwrsReport.submissionDate
            }
          />
        )}
      </Col>
    </Row>
  );
};

export default DiffDetailsContainer;
