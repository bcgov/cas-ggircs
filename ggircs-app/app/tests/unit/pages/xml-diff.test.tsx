import React from "react";
import { shallow } from "enzyme";
import XmlDiff from "pages/xml-diff";

describe("The swrs-browser page", () => {
  it("renders the diff component when it has two defined xml reports & showDiff is set to true", () => {
    const wrapper = shallow(
      <XmlDiff
        query={{
          session: null,
          allReports: { edges: null },
        }}
      />
    );
    const leftSideReport = {
      submissionDate: "Jan 1 2020",
      ecccXmlFileByEcccXmlFileId: {
        xmlFileName: "xmltestL",
        xmlFile: "<report1></report1>",
        ecccZipFileByZipFileId: {
          zipFileName: "ziptest",
        },
      },
    };
    const rightSideReport = {
      submissionDate: "Jan 2 2020",
      ecccXmlFileByEcccXmlFileId: {
        xmlFileName: "xmltestR",
        xmlFile: "<report2></report2>",
        ecccZipFileByZipFileId: {
          zipFileName: "ziptestR",
        },
      },
    };
    wrapper.setState({
      leftSideReport,
      rightSideReport,
      leftSideId: 1,
      rightSideId: 2,
      isReversed: false,
      isCollapsed: false,
      renderDiff: true,
    });
    expect(wrapper).toMatchSnapshot();
  });
});
