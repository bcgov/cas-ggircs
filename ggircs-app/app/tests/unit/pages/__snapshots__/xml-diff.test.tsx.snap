// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`The swrs-browser page renders the diff component when it has two defined xml reports & showDiff is set to true 1`] = `
<Relay(DefaultLayout)
  session={null}
  title="ECCC SWRS XML Diff"
  width="wide"
>
  <Row
    noGutters={false}
  >
    <Col
      md={
        Object {
          "order": 1,
          "span": 6,
        }
      }
    >
      <ReportSelector
        allReports={null}
        diffSide="First"
        setSwrsReportId={[Function]}
        swrsReportId={1}
      />
    </Col>
    <Col
      md={
        Object {
          "order": 2,
          "span": 6,
        }
      }
    >
      <ReportSelector
        allReports={null}
        diffSide="Second"
        setSwrsReportId={[Function]}
        swrsReportId={2}
      />
    </Col>
  </Row>
  <Row
    noGutters={false}
    style={
      Object {
        "marginTop": "2em",
      }
    }
  >
    <Col
      md={
        Object {
          "order": 1,
          "span": 6,
        }
      }
    >
      <DiffDetails
        submissionDate="Jan 1 2020"
        xmlFileName="xmltestL"
        zipFileName="ziptest"
      />
    </Col>
    <Col
      md={
        Object {
          "order": 2,
          "span": 6,
        }
      }
    >
      <DiffDetails
        submissionDate="Jan 2 2020"
        xmlFileName="xmltestR"
        zipFileName="ziptestR"
      />
    </Col>
  </Row>
  <Row
    noGutters={false}
    style={
      Object {
        "marginBottom": "2em",
        "marginTop": "2em",
      }
    }
  >
    <Col
      md={
        Object {
          "offset": 6,
          "span": 3,
        }
      }
    >
      <Row
        noGutters={false}
        style={
          Object {
            "float": "right",
          }
        }
      >
        <p
          style={
            Object {
              "marginBottom": "auto",
              "marginTop": "auto",
            }
          }
        >
          <strong>
            Summarize Changes: 
          </strong>
        </p>
        <ForwardRef(LoadableComponent)
          checked={false}
          offlabel="Off"
          onChange={[Function]}
          onlabel="On"
        />
      </Row>
    </Col>
    <Col
      md={
        Object {
          "span": 3,
        }
      }
    >
      <Button
        onClick={[Function]}
        size="small"
      >
        <FontAwesomeIcon
          border={false}
          className=""
          fixedWidth={false}
          flip={null}
          icon={
            Object {
              "icon": Array [
                512,
                512,
                Array [],
                "f362",
                "M0 168v-16c0-13.255 10.745-24 24-24h360V80c0-21.367 25.899-32.042 40.971-16.971l80 80c9.372 9.373 9.372 24.569 0 33.941l-80 80C409.956 271.982 384 261.456 384 240v-48H24c-13.255 0-24-10.745-24-24zm488 152H128v-48c0-21.314-25.862-32.08-40.971-16.971l-80 80c-9.372 9.373-9.372 24.569 0 33.941l80 80C102.057 463.997 128 453.437 128 432v-48h360c13.255 0 24-10.745 24-24v-16c0-13.255-10.745-24-24-24z",
              ],
              "iconName": "exchange-alt",
              "prefix": "fas",
            }
          }
          inverse={false}
          listItem={false}
          mask={null}
          pull={null}
          pulse={false}
          rotation={null}
          size={null}
          spin={false}
          swapOpacity={false}
          symbol={false}
          title=""
          transform={null}
        />
         Swap left/right
      </Button>
    </Col>
  </Row>
  <div
    style={
      Object {
        "height": "40em",
        "overflow": "scroll",
      }
    }
  >
    <RenderDiff
      collapse={false}
      newText="<report2></report2>"
      oldText="<report1></report1>"
    />
  </div>
</Relay(DefaultLayout)>
`;
